### PR TITLE
Handle rate limits in registration flow.

### PR DESCRIPTION
**Pull request by charlesmchen**
PTAL @michaelkirk

### CODE REVIEW

File: Signal/src/ViewControllers/CodeVerificationViewController.m [hunk 1]
```diff
@@ -302,29 +302,18 @@ - (void)verificationWasCompleted
 
 - (void)presentAlertWithVerificationError:(NSError *)error
 {
-    UIAlertController *alertController;
```
Comment (charlesmchen) [inline Signal/src/ViewControllers/CodeVerificationViewController.m]: I now convert these errors in `TSAccountManager`, so this logic can be simplified.

File: SignalServiceKit/src/Account/TSAccountManager.m [hunk 2]
```diff
@@ -386,24 +386,37 @@ - (void)verifyAccountWithCode:(NSString *)verificationCode
...
                     failureBlock(userError);
                     break;
                 }
                 case 423: {
                     DDLogError(@"%@ 2FA PIN required: %ld", self.logTag, error.code);
-                    NSError *error = OWSErrorWithCodeDescription(
-                        OWSErrorCodeRegistrationMissing2FAPIN, @"Registration missing 2FA PIN.");
+                    NSError *error = OWSErrorWithCodeDescription(OWSErrorCodeRegistrationMissing2FAPIN,
+                        NSLocalizedString(@"REGISTRATION_VERIFICATION_FAILED_WRONG_PIN",
+                            "Error message indicating that regitration failed due to a missing or incorrect 2FA PIN."));
```
Comment (charlesmchen) [inline SignalServiceKit/src/Account/TSAccountManager.m]: We should l10n these errors now that they're being surfaced to the user.

File: SignalServiceKit/src/Account/TSAccountManager.m [hunk 3]
```diff
@@ -386,24 +386,37 @@ - (void)verifyAccountWithCode:(NSString *)verificationCode
...
+                            "Error message indicating that regitration failed due to a missing or incorrect "
+                            "verification code."));
+                    failureBlock(userError);
+                    break;
+                }
+                case 413: {
+                    // In the case of the "rate limiting" error, we want to show the
+                    // "recovery suggestion", not the error's "description."
+                    NSError *userError
+                        = OWSErrorWithCodeDescription(OWSErrorCodeUserError, error.localizedRecoverySuggestion);
```
Comment (charlesmchen) [inline SignalServiceKit/src/Account/TSAccountManager.m]: Convert the 413 "rate limit" error.

File: SignalServiceKit/src/Account/TSAccountManager.m [hunk 4]
```diff
@@ -386,24 +386,37 @@ - (void)verifyAccountWithCode:(NSString *)verificationCode
...
+            OWSAssert([error.domain isEqualToString:TSNetworkManagerDomain]);
+
             DDLogWarn(@"%@ Error verifying code: %@", self.logTag, error.debugDescription);
+
             switch (error.code) {
                 case 403: {
                     NSError *userError = OWSErrorWithCodeDescription(OWSErrorCodeUserError,
                         NSLocalizedString(@"REGISTRATION_VERIFICATION_FAILED_WRONG_CODE_DESCRIPTION",
-                            "Alert body, during registration"));
+                            "Error message indicating that regitration failed due to a missing or incorrect "
```
Comment (charlesmchen) [inline SignalServiceKit/src/Account/TSAccountManager.m]: Improved comment.

File: SignalServiceKit/src/Account/TSAccountManager.m [hunk 5]
```diff
@@ -386,24 +386,37 @@ - (void)verifyAccountWithCode:(NSString *)verificationCode
...
+                    NSError *userError
+                        = OWSErrorWithCodeDescription(OWSErrorCodeUserError, error.localizedRecoverySuggestion);
                     failureBlock(userError);
                     break;
                 }
                 case 423: {
                     DDLogError(@"%@ 2FA PIN required: %ld", self.logTag, error.code);
-                    NSError *error = OWSErrorWithCodeDescription(
-                        OWSErrorCodeRegistrationMissing2FAPIN, @"Registration missing 2FA PIN.");
+                    NSError *error = OWSErrorWithCodeDescription(OWSErrorCodeRegistrationMissing2FAPIN,
```
Comment (michaelkirk) [inline SignalServiceKit/src/Account/TSAccountManager.m]: heh, I fixed this in #3101, but happy to revert my change and rebase this in.

Comment (charlesmchen) [reply SignalServiceKit/src/Account/TSAccountManager.m]: Great, I'll rebase when you've merged #3101.

Comment (charlesmchen) [reply SignalServiceKit/src/Account/TSAccountManager.m]: I just went ahead and merged - I'lll let you reconcile.   I figured it would be easier for you.

**Pull request merged**